src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/ShardedHttpExecutionServiceCluster.scala

```diff
@@ -0,0 +1,115 @@
+      .getOrElse(throw new RawlsException(s"cannot initialize ShardedHttpExecutionServiceCluster: member with default key $default does not exist"))
+
+
+  // ====================
+  // facade methods
+  // ====================
+
+  // only used in tests??? TODO: DA update tests to use the same thing runtime does, or remove this.
+  def submitWorkflow(wdl: String, inputs: String, options: Option[String], userInfo: UserInfo): Future[ExecutionServiceStatus] =
+    getDefaultMember.submitWorkflow(wdl, inputs, options, userInfo)
+
+  // by nature, this is only called for workflows that have not yet been submitted.
+  // therefore, we want to send the workflows to the cromwell instance chosen
+  // by the routing algorithm defined in this class.
+  def submitWorkflows(workflowRecs: Seq[WorkflowRecord], wdl: String, inputs: Seq[String], options: Option[String], userInfo: UserInfo): Future[(ExecutionServiceId, Seq[Either[ExecutionServiceStatus, ExecutionServiceFailure]])] = {
+    val targetMember = nextAvailableMember(workflowRecs)
+    targetMember.dao.submitWorkflows(wdl, inputs, options, userInfo) map {results =>
+      (targetMember.key, results)
+    }
+  }
+
+  // following are called on a workflow that has already been submitted.
+  // therefore, we want to use the cromwell instance that has been persisted
+  // onto that workflow.
+  def status(wfe: WorkflowExecution, userInfo: UserInfo): Future[ExecutionServiceStatus] =
+    getMember(wfe).flatMap {memb => memb.status(wfe.id, userInfo)}
+
+  def callLevelMetadata(wfe: WorkflowExecution, userInfo: UserInfo): Future[ExecutionMetadata] =
+    getMember(wfe).flatMap {memb => memb.callLevelMetadata(wfe.id, userInfo)}
+
+  def outputs(wfe: WorkflowExecution, userInfo: UserInfo): Future[ExecutionServiceOutputs] =
+    getMember(wfe).flatMap {memb => memb.outputs(wfe.id, userInfo)}
+
+  def logs(wfe: WorkflowExecution, userInfo: UserInfo): Future[ExecutionServiceLogs] =
+    getMember(wfe).flatMap {memb => memb.logs(wfe.id, userInfo)}
+
+  def abort(wfe: WorkflowExecution, userInfo: UserInfo): Future[Try[ExecutionServiceStatus]] =
+    getMember(wfe).flatMap {memb => memb.abort(wfe.id, userInfo)}
+
+
+  // ====================
+  // facade-to-cluster entry points
+  // ====================
+  // for an already-submitted workflow, get the instance to which it was submitted
+  private def getMember(wfe: WorkflowExecution):Future[ExecutionServiceDAO] = wfe.executionServiceId match {
+    case Some(execId) => Future.successful(getMemberById(ExecutionServiceId(execId)))
+    case None => {
+      dataSource.inTransaction { dataAccess =>
+        dataAccess.workflowQuery.getExecutionServiceKey(wfe.id) map {execIdOption =>
+          execIdOption match {
+            case Some(execId) => getMemberById(ExecutionServiceId(execId))
+            case None => getDefaultMember //throw new RawlsException("can only process Workflow objects with an execution service key")
+          }
+        }
+      }
+    }
+  }
```

Comment:
I think we need to do 3. It does not look that intrusive. I count 4 places that use WorkflowExecution.apply(wf: Workflow) in production code, and 1 place in ExecutionServiceClusterTest.

The reason why this is important is because this has the potential to trigger many database lookups each in their own nested transaction. For example, aborting a submission with 1000 workflows will require 1000 lookups each in their own transaction. Given our current use of slick, nested transaction can lead to deadlocks.